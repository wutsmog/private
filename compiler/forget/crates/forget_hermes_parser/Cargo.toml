[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }