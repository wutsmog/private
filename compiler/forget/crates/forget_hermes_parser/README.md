# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.