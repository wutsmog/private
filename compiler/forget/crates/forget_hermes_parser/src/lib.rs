mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
