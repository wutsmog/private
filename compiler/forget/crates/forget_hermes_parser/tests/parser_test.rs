use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
