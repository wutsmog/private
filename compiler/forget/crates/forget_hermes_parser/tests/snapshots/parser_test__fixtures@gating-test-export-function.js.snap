---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/gating-test-export-function.js
---
Input:
// @gating @forgetDirective
export function Bar(props) {
  "use forget";
  return <div>{props.bar}</div>;
}

export function NoForget(props) {
  return <Bar>{props.noForget}</Bar>;
}

export function Foo(props) {
  "use forget";
  return <Foo>{props.bar}</Foo>;
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "ExportNamedDeclaration",
      "declaration": {
        "type": "FunctionDeclaration",
        "id": {
          "type": "Identifier",
          "name": "Bar",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 0,
            "end": 1
          }
        },
        "params": [
          {
            "type": "Identifier",
            "name": "props",
            "typeAnnotation": null,
            "loc": null,
            "range": {
              "start": 0,
              "end": 1
            }
          }
        ],
        "body": {
          "type": "BlockStatement",
          "body": [
            {
              "type": "ExpressionStatement",
              "expression": {
                "type": "StringLiteral",
                "value": "use forget",
                "loc": null,
                "range": {
                  "start": 0,
                  "end": 1
                }
              },
              "directive": null,
              "loc": null,
              "range": {
                "start": 0,
                "end": 1
              }
            },
            {
              "type": "ReturnStatement",
              "argument": {
                "type": "JSXElement",
                "openingElement": {
                  "type": "JSXOpeningElement",
                  "name": {
                    "type": "JSXIdentifier",
                    "name": "div",
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  },
                  "attributes": [],
                  "selfClosing": false,
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "children": [
                  {
                    "type": "JSXExpressionContainer",
                    "expression": {
                      "type": "MemberExpression",
                      "object": {
                        "type": "Identifier",
                        "name": "props",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      },
                      "property": {
                        "type": "Identifier",
                        "name": "bar",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      },
                      "computed": false,
                      "loc": null,
                      "range": {
                        "start": 0,
                        "end": 1
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  }
                ],
                "closingElement": {
                  "type": "JSXClosingElement",
                  "name": {
                    "type": "JSXIdentifier",
                    "name": "div",
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  },
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "loc": null,
                "range": {
                  "start": 0,
                  "end": 1
                }
              },
              "loc": null,
              "range": {
                "start": 0,
                "end": 1
              }
            }
          ],
          "loc": null,
          "range": {
            "start": 0,
            "end": 1
          }
        },
        "generator": false,
        "async": false,
        "loc": null,
        "range": {
          "start": 0,
          "end": 1
        },
        "loc": null,
        "range": {
          "start": 0,
          "end": 1
        }
      },
      "specifiers": [],
      "source": null,
      "loc": null,
      "range": {
        "start": 0,
        "end": 1
      }
    },
    {
      "type": "ExportNamedDeclaration",
      "declaration": {
        "type": "FunctionDeclaration",
        "id": {
          "type": "Identifier",
          "name": "NoForget",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 0,
            "end": 1
          }
        },
        "params": [
          {
            "type": "Identifier",
            "name": "props",
            "typeAnnotation": null,
            "loc": null,
            "range": {
              "start": 0,
              "end": 1
            }
          }
        ],
        "body": {
          "type": "BlockStatement",
          "body": [
            {
              "type": "ReturnStatement",
              "argument": {
                "type": "JSXElement",
                "openingElement": {
                  "type": "JSXOpeningElement",
                  "name": {
                    "type": "JSXIdentifier",
                    "name": "Bar",
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  },
                  "attributes": [],
                  "selfClosing": false,
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "children": [
                  {
                    "type": "JSXExpressionContainer",
                    "expression": {
                      "type": "MemberExpression",
                      "object": {
                        "type": "Identifier",
                        "name": "props",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      },
                      "property": {
                        "type": "Identifier",
                        "name": "noForget",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      },
                      "computed": false,
                      "loc": null,
                      "range": {
                        "start": 0,
                        "end": 1
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  }
                ],
                "closingElement": {
                  "type": "JSXClosingElement",
                  "name": {
                    "type": "JSXIdentifier",
                    "name": "Bar",
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  },
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "loc": null,
                "range": {
                  "start": 0,
                  "end": 1
                }
              },
              "loc": null,
              "range": {
                "start": 0,
                "end": 1
              }
            }
          ],
          "loc": null,
          "range": {
            "start": 0,
            "end": 1
          }
        },
        "generator": false,
        "async": false,
        "loc": null,
        "range": {
          "start": 0,
          "end": 1
        },
        "loc": null,
        "range": {
          "start": 0,
          "end": 1
        }
      },
      "specifiers": [],
      "source": null,
      "loc": null,
      "range": {
        "start": 0,
        "end": 1
      }
    },
    {
      "type": "ExportNamedDeclaration",
      "declaration": {
        "type": "FunctionDeclaration",
        "id": {
          "type": "Identifier",
          "name": "Foo",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 0,
            "end": 1
          }
        },
        "params": [
          {
            "type": "Identifier",
            "name": "props",
            "typeAnnotation": null,
            "loc": null,
            "range": {
              "start": 0,
              "end": 1
            }
          }
        ],
        "body": {
          "type": "BlockStatement",
          "body": [
            {
              "type": "ExpressionStatement",
              "expression": {
                "type": "StringLiteral",
                "value": "use forget",
                "loc": null,
                "range": {
                  "start": 0,
                  "end": 1
                }
              },
              "directive": null,
              "loc": null,
              "range": {
                "start": 0,
                "end": 1
              }
            },
            {
              "type": "ReturnStatement",
              "argument": {
                "type": "JSXElement",
                "openingElement": {
                  "type": "JSXOpeningElement",
                  "name": {
                    "type": "JSXIdentifier",
                    "name": "Foo",
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  },
                  "attributes": [],
                  "selfClosing": false,
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "children": [
                  {
                    "type": "JSXExpressionContainer",
                    "expression": {
                      "type": "MemberExpression",
                      "object": {
                        "type": "Identifier",
                        "name": "props",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      },
                      "property": {
                        "type": "Identifier",
                        "name": "bar",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      },
                      "computed": false,
                      "loc": null,
                      "range": {
                        "start": 0,
                        "end": 1
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  }
                ],
                "closingElement": {
                  "type": "JSXClosingElement",
                  "name": {
                    "type": "JSXIdentifier",
                    "name": "Foo",
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  },
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "loc": null,
                "range": {
                  "start": 0,
                  "end": 1
                }
              },
              "loc": null,
              "range": {
                "start": 0,
                "end": 1
              }
            }
          ],
          "loc": null,
          "range": {
            "start": 0,
            "end": 1
          }
        },
        "generator": false,
        "async": false,
        "loc": null,
        "range": {
          "start": 0,
          "end": 1
        },
        "loc": null,
        "range": {
          "start": 0,
          "end": 1
        }
      },
      "specifiers": [],
      "source": null,
      "loc": null,
      "range": {
        "start": 0,
        "end": 1
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 0,
    "end": 1
  }
}
